src/practitest_firecracker/practitest.clj

```diff
@@ -172,27 +189,65 @@
                                       :attributes {:set-id  testset-id
                                                    :test-id test-id}}}}))))
 
-(defn ll-create-run [{:keys [base-uri credentials max-api-rate-throttler]} project-id instance-id attributes steps]
+(defn has-duplicates? [key runs]
+  (let [grouped (group-by key (into [] runs))]
+    (not (= (count runs) (count grouped)))))
+
+(defn ll-create-instances [{:keys [base-uri credentials max-api-rate-throttler]} [project-id display-action-logs] instances]
+  (when display-action-logs (log/infof "create instances"))
+  (let [uri (build-uri base-uri create-instance-uri project-id)]
+    (api-call {:credentials  credentials
+               :uri          uri
+               :method       (max-api-rate-throttler http/post)
+               :form-params  {:data instances}})))
+
+(defn ll-create-run [{:keys [base-uri credentials max-api-rate-throttler]} [project-id display-action-logs] instance-id attributes steps]
+  (when display-action-logs (log/infof "create run for instance %s" instance-id))
   (let [uri (build-uri base-uri create-run-uri project-id)]
     (first
      (api-call {:credentials  credentials
                 :uri          uri
                 :method       (max-api-rate-throttler http/post)
                 :form-params  {:data {:type       "instances"
-                                      :attributes (assoc attributes :instance-id instance-id)
+                                      :attributes (merge attributes {:instance-id instance-id})
                                       :steps      {:data steps}}}}))))
 
-(defn ll-create-runs [{:keys [base-uri credentials max-api-rate-throttler]} project-id runs]
+(defn ll-create-runs [{:keys [base-uri credentials max-api-rate-throttler] :as client} [project-id display-action-logs] runs]
+  (when display-action-logs (log/infof "create runs"))
   (let [uri (build-uri base-uri create-run-uri project-id)]
+    (if (has-duplicates? :instance-id runs)
+      (let [grouped    (group-by :instance-id runs)
+            duplicates (map first (vals (into {} (filter #(> (count (last %)) 1) grouped))))
+            uniq       (map first (vals (into {} (filter #(= (count (last %)) 1) grouped))))]
+        (do (doall
```

Comment:
why do you need `do` here? It's already inside `let`

Reply:
do is from debug logs that I had to add do to print stuff, will clean it